app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/WifiDirectManager.kt

```diff
@@ -239,49 +214,64 @@ class WifiDirectManager @Inject constructor(
     })
   }
 
-  private fun performHandshakeWithSelectedPeerDevice() {
+  private fun performHandshakeWith(groupInfo: WifiP2pInfo) {
     if (BuildConfig.DEBUG) {
       Log.d(TAG, "Starting handshake")
     }
-    peerGroupHandshakeAsyncTask = PeerGroupHandshakeAsyncTask(this)
-      .also { it.execute() }
+    lifecycleCoroutineScope.launch {
+      val peerGroupHandshake = if (isFileSender) {
+        SenderHandShake(this@WifiDirectManager, groupInfo)
+      } else {
+        ReceiverHandShake(this@WifiDirectManager, groupInfo)
+      }
+      val inetAddress = peerGroupHandshake.handshake()
+      if (inetAddress != null) {
+        startFileTransfer(groupInfo, inetAddress)
+      } else {
+        if (BuildConfig.DEBUG) {
+          Log.d(TAG, "InetAddress is null")
+        }
+        onFileTransferAsyncTaskComplete(false)
+        activity.toast(R.string.connection_refused)
+      }
+    }
   }
 
   val totalFilesForTransfer: Int
     get() = filesForTransfer.size
 
   fun getFilesForTransfer() = filesForTransfer
 
-  fun setFilesForTransfer(fileItems: ArrayList<FileItem>) {
+  fun setFilesForTransfer(fileItems: List<FileItem>) {
     filesForTransfer = fileItems
   }
 
-  val zimStorageRootPath
-    get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
-
-  fun getFileReceiverDeviceAddress() = fileReceiverDeviceAddress
-
-  fun setClientAddress(clientAddress: InetAddress) {
-
-    // If control reaches here, means handshake was successful
-    selectedPeerDeviceInetAddress = clientAddress
-    startFileTransfer()
-  }
+  val zimStorageRootPath get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
 
-  private fun startFileTransfer() {
-    if (isGroupFormed) {
+  private fun startFileTransfer(groupInfo: WifiP2pInfo, inetAddress: InetAddress) {
+    if (groupInfo.groupFormed) {
       if (isFileSender) {
         Log.d(LocalFileTransferFragment.TAG, "Starting file transfer")
-        fileReceiverDeviceAddress =
-          if (isGroupOwner) selectedPeerDeviceInetAddress else groupOwnerAddress
+        val fileReceiverDeviceAddress =
+          if (groupInfo.isGroupOwner) inetAddress else groupInfo.groupOwnerAddress
         activity.toast(R.string.preparing_files, Toast.LENGTH_LONG)
-        senderDeviceAsyncTask = SenderDeviceAsyncTask(this, activity).also {
-          it.execute(*filesForTransfer.toTypedArray())
+        val senderDevice = SenderDevice(activity, this, fileReceiverDeviceAddress)
+        lifecycleCoroutineScope.launch {
```

Comment:
we actually don't need to call this again, we are in the same functional scope as the original `launch` call. make `startFileTransfer` `suspend` and it should all be good

Reply:
D'oh! I forget about they are calling from the same scope!